onnxscript/codeanalysis/onnx_to_onnxscript.py

```diff
@@ -0,0 +1,533 @@
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+
+from __future__ import annotations
+
+import re
+from dataclasses import dataclass
+from pathlib import Path
+from typing import BinaryIO, Final, Literal, Sequence, cast, overload
+
+import libcst as cst
+import libcst.matchers as cstm
+import libcst.metadata as cstmeta
+import onnx
+
+from onnxscript.codeanalysis import (
+    CstCodeGenerator,
+    RemoveUnusedImportsTransformer,
+    format_code,
+    make_const_expr,
+)
+
+__all__ = [
+    "OnnxScriptCodeGenerator",
+    "OnnxScriptTransformer",
+    "OnnxToPythonOperatorTransformer",
+    "OnnxConstantOpToPythonConstantTransformer",
+    "Driver",
+]
+
+DEFAULT_OPSET_VERSION: Final = 18
+
+
+@dataclass
+class QualifiedOnnxOp:
+    domain: str
+    name: str
+    version: int = 0
+
+    def __eq__(self, __value: object) -> bool:
+        if isinstance(__value, str):
+            return self.domain == "" and self.name == __value
+        elif isinstance(__value, QualifiedOnnxOp):
+            return (
+                self.domain == __value.domain
+                and self.name == __value.name
+                and (__value.version <= 0 or self.version == __value.version)
+            )
+        return False
+
+
+class OnnxScriptCodeGenerator(CstCodeGenerator):
+    def __init__(self):
+        super().__init__()
+        self.__opset_version = DEFAULT_OPSET_VERSION
+
+    def translate_model_proto(self, model_proto: onnx.ModelProto) -> cst.Module:
+        for opset_import in model_proto.opset_import:
+            if opset_import.domain == "":
+                self.__opset_version = opset_import.version
+
+        self.add_import_from("__future__", "annotations")
+
+        model_local_functions = [
+            self.translate_function_proto(func) for func in model_proto.functions
+        ]
+        model_main_function = self.__make_function(model_proto.graph, "script")
+
+        return cst.Module(
+            body=[
+                *self.make_import_statements(),
+                model_main_function,
+                *model_local_functions,
+            ]
+        )
+
+    def translate_function_proto(self, function_proto: onnx.FunctionProto) -> cst.FunctionDef:
+        return self.__make_function(function_proto, "script")
+
+    def translate_graph_proto(
+        self,
+        function_proto: onnx.GraphProto,
+        func_type: Literal["graph"] | Literal["script"] = "graph",
+    ) -> cst.FunctionDef:
+        return self.__make_function(function_proto, func_type)
+
+    def translate_tensor_proto(self, tensor_proto: onnx.TensorProto) -> cst.BaseExpression:
+        if onnx.external_data_helper.uses_external_data(tensor_proto):
+            raise NotImplementedError("tensors with external data are not supported")
+
+        self.add_import("onnx")
+
+        numpy_tensor = onnx.numpy_helper.to_array(tensor_proto)
+
+        return cst.Call(
```

Comment:
This is one example where verbosity makes this harder to read than the existing template-based mechanism. Do you think it might make sense to judiciously mix parsed strings, with formatted strings, with cst constructors to get the best of both worlds?